ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,523 @@
+   * in a {@link AccountInfoMap}, according to the Javadoc of {@link AccountService}. Two {@link Account}s can be
+   * conflicting with each other if they have different account Ids but the same account name.
+   *
+   * @param accountsToSet The collection of {@link Account}s to check conflict.
+   * @param currentAccounts A {@link AccountInfoMap} that represents a group of {@link Account}s to check conflict.
+   * @return {@code true} if there is at least one {@link Account} in {@code accountPairs} conflicts with the existing
+   *                      {@link Account}s in {@link AccountInfoMap}, {@code false} otherwise.
+   */
+  private boolean hasConflictingAccount(Collection<Account> accountsToSet, AccountInfoMap currentAccounts) {
+    for (Account account : accountsToSet) {
+      Account remoteAccount = currentAccounts.getAccountById(account.getId());
+
+      // @todo if remoteAccount is not null, make sure
+      // @todo account.generationId() = accountInfoMap.getById(account.getId()).generationId()+1
+      // @todo the generationId will be added to Account class.
+
+      // check for case D in the javadoc of AccountService
+      if (remoteAccount == null && currentAccounts.containsName(account.getName())) {
+        Account conflictingAccount = currentAccounts.getAccountByName(account.getName());
+        logger.info(
+            "Account to update with accountId={} accountName={} conflicts with an existing record with accountId={} accountName={}",
+            account.getId(), account.getName(), conflictingAccount.getId(), conflictingAccount.getName());
+        return true;
+      }
+
+      // check for case E in the javadoc of AccountService
+      if (remoteAccount != null && !account.getName().equals(remoteAccount.getName()) && currentAccounts.containsName(
+          account.getName())) {
+        Account conflictingAccount = currentAccounts.getAccountByName(account.getName());
+        logger.info(
+            "Account to update with accountId={} accountName={} conflicts with an existing record with accountId={} accountName={}",
+            account.getId(), account.getName(), conflictingAccount.getId(), conflictingAccount.getName());
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks if the {@code HelixAccountService} is open.
+   */
+  private void checkOpen() {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+  }
+
+  /**
+   * <p>
+   *   A cache to store {@link Account} information locally. It implements the {@link AccountService} interface,
+   *   and can respond to the queries for {@link Account}. Internally it maintains a {@link AccountInfoMap} that
+   *   represents a group of {@link Account}s that can be indexed by both account id and name.
+   * </p>
+   * <p>
+   *   Operations on a {@link Cache} is thread-safe.
+   * </p>
+   */
+  private class Cache implements AccountService {
```

Comment:
Removed, now `HelixAccountService` has an `AccountInfoMapRef` pointing to an `AccountInfoMap`.